crates/matrix-sdk-crypto/src/store/locks.rs

```diff
@@ -14,21 +14,40 @@
 
 //! Collection of small helpers that implement store-based locks.
 //!
-//! Those locks are implemented as one value in the key-value crypto store, that
-//! exists if and only if the lock has been taken. For this to work correctly,
-//! we rely on multiple assumptions:
+//! This is a per-process lock that may be used only for very specific use
+//! cases, where multiple processes might concurrently write to the same
+//! database at the same time; this would invalidate crypto store caches, so
+//! that should be done mindfully. Such a lock can be acquired multiple times by
+//! the same process, and it remains active as long as there's at least one user
+//! in a given process.
 //!
-//! - the store must allow concurrent reads and writes from multiple processes.
-//!   For instance, for
-//! sqlite, this means that it is running in [WAL](https://www.sqlite.org/wal.html) mode.
-//! - the two operations used in the store implementation,
-//!   `insert_custom_value_if_missing` and
-//! `remove_custom_value`, must be atomic / implemented in a transaction.
+//! The lock is implemented using time-based leases to values inserted in a
+//! crypto store. The store maintains the lock identifier (key), who's the
+//! current holder (value), and an expiration timestamp on the side; see also
+//! `CryptoStore::try_take_leased_lock` for more details.
+//!
+//! The lock is initially acquired for a certain period of time (namely, the
+//! duration of a lease, aka `LEASE_DURATION_MS`), and then a "heartbeat" task
+//! renews the lease to extend its duration, every so often (namely, every
+//! `EXTEND_LEASE_EVERY_MS`). Since the tokio scheduler might be busy, the
+//! extension request should happen way more frequently than the duration of a
+//! lease, in case a deadline is missed. The current values have been chosen to
+//! reflect that, with a ratio of 1:10 as of 2023-06-23.
+//!
+//! Releasing the lock happens naturally, by not renewing a lease. It happens
+//! automatically after the duration of the last lease, at most.
+//!
+//! For this to work, the store must allow concurrent reads and writes from
+//! multiple processes. For instance, for sqlite, this means that it is running in [WAL](https://www.sqlite.org/wal.html)
+//! mode.
```

Comment:
I don't understand this. Why is this necessary?

Reply:
Yup, I don't think WAL mode is relevant here. You can still have multi-process support without WAL, but it'll block on the SQLite side.

Reply:
Ah, cool; I misunderstood the role of WAL then. Removed that paragraph from the comment.